### ISSUE TITLE
CentOS 8.4 rootless podman: [nvidia-container-cli: initialization error: driver error: failed to process request]

### ISSUE THREAD

**Issue by SoloGao** [NVIDIA/nvidia-docker#1565]
### 1. Issue or feature description

Hi all, I just installed Nvidia container toolkit under CentOS 8.4 with podman 3.2.3 using the official guide: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#podman . The driver is working properly (nvidia-smi shows all GPUs) and podman with root also works (no-cgroups = false). However, when running rootless with:
 - no-cgroups = true
 - subuid & subgid set
 - all directories in local disk with proper permissions (not on NFS server)
 - podman run --rm --security-opt=label=disable  --hooks-dir=/usr/share/containers/oci/hooks.d/ nvcr.io/nvidia/cuda nvidia-smi

I got an error:

> Error: OCI runtime error: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: initialization error: driver error: failed to process request

And the `nvidia-container-toolkit.log` may tell why:

> -- WARNING, the following logs are for debugging purposes only --
> 
> I1110 14:05:03.689985 15116 nvc.c:372] initializing library context (version=1.5.1, build=4afad130c4c253abd3b2db563ffe9331594bda41)
> I1110 14:05:03.690118 15116 nvc.c:346] using root /
> I1110 14:05:03.690126 15116 nvc.c:347] using ldcache /etc/ld.so.cache
> I1110 14:05:03.690134 15116 nvc.c:348] using unprivileged user 65534:65534
> I1110 14:05:03.690157 15116 nvc.c:389] attempting to load dxcore to see if we are running under Windows Subsystem for Linux (WSL)
> I1110 14:05:03.690295 15116 nvc.c:391] dxcore initialization failed, continuing assuming a non-WSL environment
> W1110 14:05:03.690316 15116 nvc.c:249] skipping kernel modules load due to user namespace
> I1110 14:05:03.690608 15121 driver.c:101] starting driver service
> **E1110 14:05:03.690731 15121 driver.c:168] could not start driver service: privilege change failed: invalid argument**
> I1110 14:05:03.695092 15116 driver.c:203] driver service terminated successfully

Has anyone know how to deal with it? Thanks!

### 2. Steps to reproduce the issue

 - clean install a CentOS 8
 - install nvidia driver using https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html#centos8
 - install nvidia-container-toolkit using https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#podman
 - set rootless by adding subuid and subgid to the user account/change the path of the registery/set no-cgroups to true
 - start rootless docker with nvidia-container-toolkit

**Comment by SoloGao** [comment]
I found the problem and came up with this solution. Just for anyone who googled this.

The major reason is that podman needs subuid/subgid, and need the number of subuids to be more than the max uid inside an image. Therefore, I just assigned 10000 subuids/subgids for each user.

However, for nvidia-container-toolkit, it inits the user namespace with unprivileged user 65534:65534. Therefore, if you assign subuids/subgids less than 65535, the process is not able to run properly.

Therefore, I just enlarge the subuid/subgid range for each user, and it works. Another way is to change `/etc/nvidia-container-runtime/config.toml`, and set the uid/gid of the "user" inside your subuid/subgid range (in my example, 9999:9999).
